src/bunit/Rendering/BunitRendererSynchronizationContextDispatcher.cs

```diff
@@ -0,0 +1,210 @@
+using System.Diagnostics;
+
+namespace Bunit.Rendering;
+
+internal class BunitRendererSynchronizationContextDispatcher : Dispatcher
+{
+	private readonly object bulkheadsLock = new();
+	private readonly BunitRendererSynchronizationContext context;
+	private readonly Queue<TaskCompletionSource> workItemBulkheads = new();
+	private Func<bool>? isReadyForDispatch;
+	private Action? onAfterDispatchCompleted;
+	private Action<Exception>? onDispatchException;
+
+	private Task CreateBulkhead()
+	{
+		lock (bulkheadsLock)
+		{
+			if (isReadyForDispatch is not null && isReadyForDispatch())
+			{
+				return Task.CompletedTask;
+			}
+			else
+			{
+				var tcs = new TaskCompletionSource();
```

Comment:
I would also here lean into the recommendation of using `RunContinuationsAsynchronously`
```suggestion
				var tcs = new TaskCompletionSource(TaskCreationOptions. RunContinuationsAsynchronously);
```

Reply:
I just made that change, and all my new tests are passing (lots of old tests are still broken though).